### ISSUE TITLE
Check compatibility with PHP 7

### ISSUE THREAD

**Issue by ezzatron** [eloquent/phony#94]
- [x] Throwable interface / engine error exceptions
- [x] Scalar type hints
- [x] Return type declarations
- [x] Anonymous classes
- [x] Closure::call()
- [x] Expectations
- [x] Generator return expressions

**Comment by jmalloc** [comment]
What needs to be done regarding assert/expectations?

**Comment by ezzatron** [comment]
@jmalloc Sorry, just saw that comment re expectations.

I wanted to see if it made sense to provide a facade / namespaced functions that use the new expectations as their underlying mechanism for recording verification failures.

This would be for standalone usage / potentially some future testing frameworks might provide integration that way. I just want to understand the possibilities.

JavaScript testing frameworks tend to be better at separating the test runner and assertions. I was hoping expectations might provide a path to something similar for PHP.

**Comment by jmalloc** [comment]
Ah, I see.

TL;DR, In PHP 7, assert() can either throw or produce a warning depending on (a) how it's called, (b) an [INI setting](http://php.net/manual/en/function.assert.php#function.assert.expectations).

Their real power is that they can be optimised out of the bytecode entirely (based on a separate INI setting), providing a true dev-time-only check.

IMO there's not a lot of use in basing a facade on `assert()`, because you'd never want it to be optimised out, and it doesn't offer anything more than an exception.

**Comment by ezzatron** [comment]
However, if I were writing a test runner, I'd probably handle [AssertionError](http://php.net/assertionerror) as a generic, always-available exception that can be distinguished from exceptions that are truly errors in the code.

It would be a small step towards not needing explicit integration for every test runner.

**Comment by jmalloc** [comment]
I'm not sure I follow. What would you do differently with an assertion error vs a "real" error? And, if you'd do anything differently at all, how would you distinguish between an assertion failure produced by the runner and one produced by the user's code?

**Comment by ezzatron** [comment]
It would be the same thing as a test *error* vs a test *failure* in PHPUnit. `AssertionError` is a failure, anything else is an error.

And I get your point re distinguishing them, but technically the same problem exists for any exception the test runner chooses as its 'assertion failed' exception. If I throw a `PHPUnit_Framework_ExpectationFailedException` in my code, PHPUnit can't distinguish that from its own internals.

Furthermore, if you are using `assert()` in user code, it may actually be a *good* thing for that to hook in to the test runner's output. That seems like the primary use-case - catching bad stuff at dev time.

**Comment by jmalloc** [comment]
Ok, I see what you mean, though by "test runner" you're referring to hypothetical-non-phony-thing? If so then yeah I'd tend to agree, though for whatever it's worth I don't think I've ever derived any value from noting the difference between `F` and `E` in PHPUnit.

The "distinguishing problem" is **far** less likely using a custom exception type, but I take *your* point ;)

Ok so.. what about `Closure::call()` anything in mind there in particular?

**Comment by ezzatron** [comment]
`Closure::call()` was mentioned in the PHP 7 new features page as being "more performant":

> Closure::call() is a more performant, shorthand way of temporarily binding an object scope to a closure and invoking it.

So I just wanted to check if we could replace any usage of `$closure->bind()` with it, but honestly, the check to see if we can use it will probably negate any performance benefits, and I'm not 100% sure we even use `$closure->bind()` at all.

It's the least important point on this task, basically.

**Comment by jmalloc** [comment]
Ok cool, that probably means that the PHP 7 stuff I can be helpful with is done until you get to the overhaul of the generator stuff. I'll probably take a look at #113 though, and #112 if I'm feeling adventurous.

**Comment by ezzatron** [comment]
Cool man, thanks for the help.

**Comment by ezzatron** [comment]
@jmalloc FYI I rebased off of the latest bug fixes & documentation update. If you do any more work on this branch, you'll probably have to blow away & re-pull it first.

**Comment by jmalloc** [comment]
No probs, I haven't switched over from my fork yet.

**Comment by ezzatron** [comment]
There's only one place in the code where Phony uses `Closure::bindTo()`, which is to call private parent constructors in generated mock code. I don't think it's worth an extra check there to use `Closure::call()` under PHP 7, so I'm marking that one done.

**Comment by ezzatron** [comment]
I think support for an expectation-based facade and functions can come later, so I'm marking that one done.

**Comment by ezzatron** [comment]
Marking generator return expressions done in favour of #115.

**Comment by ezzatron** [comment]
All tasks are now represented by other issues or PRs.

**Comment by ezzatron** [comment]
Go get your 0.7.0 now, hot off the press!
